lib/srv/alpnproxy/local_proxy_test.go

```diff
@@ -246,6 +249,50 @@ func TestMiddleware(t *testing.T) {
 	m.waitForCounts(t, 1, 1)
 }
 
+type mockMiddlewareConnUnauth struct {
+}
+
+func (m *mockMiddlewareConnUnauth) OnNewConnection(_ context.Context, _ *LocalProxy, _ net.Conn) error {
+	return trace.AccessDenied("access denied.")
+}
+
+func (m *mockMiddlewareConnUnauth) OnStart(_ context.Context, _ *LocalProxy) error {
+	return nil
+}
+
+var _ LocalProxyMiddleware = (*mockMiddlewareConnUnauth)(nil)
+
+func TestLocalProxyClosesConnOnError(t *testing.T) {
+	m := &mockMiddlewareConnUnauth{}
+	hs := httptest.NewTLSServer(http.HandlerFunc(func(writer http.ResponseWriter, request *http.Request) {}))
+	lp, err := NewLocalProxy(LocalProxyConfig{
+		Listener:           mustCreateLocalListener(t),
+		RemoteProxyAddr:    hs.Listener.Addr().String(),
+		Protocols:          []common.Protocol{common.ProtocolHTTP},
+		ParentContext:      context.Background(),
+		InsecureSkipVerify: true,
+		Middleware:         m,
+	})
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		err := lp.Close()
+		require.NoError(t, err)
+		hs.Close()
+	})
+	go func() {
+		require.NoError(t, lp.Start(context.Background()))
+	}()
+
+	conn, err := net.Dial("tcp", lp.GetAddr())
+	require.NoError(t, err)
+
+	conn.SetReadDeadline(time.Now().Add(5 * time.Second))
```

Comment:
The purpose of the read deadline is so that if the connection is not closed (by some future breaking change), the test fails within 5 seconds instead of hanging until the entire test suite times out. I would only remove it if this test ran in parallel to prevent flakiness, but it's such a quick test that I don't think it needs to be run in parallel.

I always run tests without my fixes first to verify they fail, and without the deadline it would just hang (default timeout for `go test` is 10 minutes).